best practices: Advanced Multi-Column Index Optimization (#20955)
pingcap/docs: best-practices/multi-column-index-optimization.md

---

> title: Use Multi-Column Index to Optimize Queries
> summary: Learn how to use multi-column indexes effectively in TiDB and understand advanced optimization techniques.
> ---
> 
> # Use Multi-Column Index to Optimize Queries

Author: ```suggestion
title: Multi-Column Index Optimization
summary: Learn how to use multi-column indexes effectively in TiDB and understand advanced optimization techniques.
---

# Multi-Column Index Optimization
```

> In today’s data-driven world, efficiently handling complex queries on large datasets is critical to keeping applications responsive and performant. For TiDB, a distributed SQL database designed to manage high-scale, high-demand environments, optimizing data access paths is essential to delivering smooth, efficient queries. This post dives into TiDB’s advanced approach to optimizing multi-column indexes. This is a powerful feature that significantly enhances the speed and accuracy of data retrieval across distributed clusters.

Author: ```suggestion
In today's data-driven world, efficiently handling complex queries on large datasets is critical to keeping applications responsive and performant. For TiDB, a distributed SQL database designed to manage high-scale, high-demand environments, optimizing data access paths is essential to delivering smooth, efficient queries. This post dives into TiDB's advanced approach to optimizing multi-column indexes. This is a powerful feature that significantly enhances the speed and accuracy of data retrieval across distributed clusters.
```

> TiDB’s query optimizer leverages multi-column indexes to intelligently filter data, handling complex query conditions that traditional databases like MySQL can’t process as effectively. Here, we’ll walk through how multi-column indexes function, why they’re crucial, and how TiDB’s optimization transforms intricate query conditions into efficient access paths. The impact? Faster responses, minimized table scans, and streamlined performance—even at massive scale.

Author: ```suggestion
TiDB's query optimizer leverages multi-column indexes to intelligently filter data, handling complex query conditions that traditional databases like MySQL cannot process as effectively. This document walks you through how multi-column indexes function, why they are crucial, and how TiDB's optimization transforms intricate query conditions into efficient access paths. The impact? Faster responses, minimized table scans, and streamlined performance, even at massive scale.
```

> Without these optimizations, query performance in large TiDB databases can degrade quickly. Full table scans and inadequate filtering can turn milliseconds into minutes. Additionally, excessive memory use can lead to out-of-memory (OOM) errors, especially in constrained environments. TiDB’s targeted approach ensures only relevant data is accessed. This keeps latency low and memory usage efficient—even for the most complex queries. Let’s explore how TiDB makes it happen.

Author: ```suggestion
Without these optimizations, query performance in large TiDB databases can degrade quickly. Full table scans and inadequate filtering can turn milliseconds into minutes. Additionally, excessive memory use can lead to out-of-memory (OOM) errors, especially in constrained environments. TiDB's targeted approach ensures only relevant data is accessed. This keeps latency low and memory usage efficient, even for the most complex queries. This document explores how TiDB makes it happen.
```

## Background: Multi-Column Indexes

> Indexes are a powerful tool for improving query performance by avoiding the need to scan all rows in a table. Let’s take an example of a rental listings table defined as follows. In this example, each listing contains a unique ID, city, number of bedrooms, rent price, and availability date:

Author: ```suggestion
Indexes are a powerful tool for improving query performance by avoiding the need to scan all rows in a table. This document takes an example of a rental listings table defined as follows. In this example, each listing contains a unique ID, city, number of bedrooms, rent price, and availability date:
```

);
```

Suppose this table has 20 million listings across the United States. If we want to find all listings with a price under $2,000, we could add an index on the price column. This index would allow the optimizer to filter out rows, scanning only the range [ -inf, 2000.00 ). This helps reduce the search to about 14 million rows (assuming 70% of rentals are priced above $2,000). In the query execution plan, TiDB would perform an index range scan on price. This limits the need for a full table scan and improves efficiency.

```sql
"Q1" 
explain FORMAT=BRIEF
    SELECT * FROM listings WHERE price < 2000;
-----+------------------------------------------------------+--------------------------
| id  task                | access object                        | operator info   
+-----------------------------+---------+-----------+-----------------------------------
| IndexLookUp             | root                                 | 
| ├─IndexRangeScan(Build) | table:listings,index:price_idx(price)| range:[-inf,2000.00) 
| └─TableRowIDScan(Probe) | table:listings                       | 
+-----------------------------+---------+-----------+-----------------------------------
```

> While this filter improves performance, it may still return a large number of rows. This isn’t ideal for a user looking for more specific listings. Adding filters, such as specifying the city, number of bedrooms, and a maximum price, narrows the results significantly. For example, a query to find two-bedroom listings in San Francisco under $2,000 is more useful, likely returning only a few dozen rows.

Author: ```suggestion
While this filter improves performance, it might still return a large number of rows. This is not ideal for a user looking for more specific listings. Adding filters, such as specifying the city, number of bedrooms, and a maximum price, narrows the results significantly. For example, a query to find two-bedroom listings in San Francisco under $2,000 is more useful, likely returning only a few dozen rows.
```

To optimize this query, we can create a multi-column index on `city`, `bedrooms`, and `price`:

```sql
CREATE INDEX idx_city_bedrooms_price ON listings (city, bedrooms, price);
```

Multi-column indexes in SQL are lexicographically ordered, meaning they are sorted first by `city`, then by `bedrooms` within each city, and finally by `price` within each city-bedroom combination. This ordering allows TiDB to efficiently access rows based on each condition:

1. Filtering by `city` as the primary filter.
2. Optionally filtering by `bedrooms` within that city.
3. Optionally filtering by `price` within the city-bedroom grouping.

## Sample Data

> Let’s look at a sample dataset to illustrate how multi-column indexing refines search results:

Author: ```suggestion
Look at a sample dataset to illustrate how multi-column indexing refines search results:
```

| City | Bedrooms | Price |
| --- | --- | --- |
| San Diego | 1 | 1000 |
| San Diego | 1 | 1500 |
| San Diego | 2 | 1000 |
| San Diego | 2 | 2500 |
| San Diego | 3 | 1000 |
| San Diego | 3 | 2500 |
| San Francisco | 1 | 1000 |
| San Francisco | 1 | 1500 |
| San Francisco | 2 | 1000 |
| San Francisco | 2 | 1500 |
| San Francisco | 3 | 2500 |
| San Francisco | 3 | 3000 |

## Optimized queries and results

Using the multi-column index, TiDB can efficiently narrow the range to find listings in San Francisco with two bedrooms and a price under `$2,000`:

```sql
-- Query 2: Find two-bedroom listings in San Francisco under $2,000
EXPLAIN FORMAT=BRIEF 
    SELECT * FROM listings 
    WHERE city = 'San Francisco' AND bedrooms = 2 AND price < 2000;
```

```sql
-- Plan for Query 2
EXPLAIN FORMAT=BRIEF 
    SELECT * FROM listings 
    WHERE city = 'San Francisco' and bedrooms = 2 and price < 2000;

> -----+------------------------------------------------------+--------------------------

Author: Fixed

                          | index:idx_city_bedrooms_price| ["New York" 2 1000.00,
                          | (city, bedrooms, price)      |  "New York" 2 2500.00)
| └─TableRowIDScan(Probe) | table:listings               | 
+-----------------------------+---------+-----------+-----------------------------------
```

### Example 2: non-overlapping ranges

In a different scenario, imagine a query that looks for affordable single-bedroom listings in either San Francisco or San Diego. Here, the `OR` condition specifies two distinct ranges for different cities:

- Listings in San Francisco, 1 bedroom, priced between `$1,500` and `$2,500`
- Listings in San Diego, 1 bedroom, priced between `$1,000` and `$1,500`

Because there is no overlap between these ranges, they remain separate in the execution plan, with each city having its own index range:

```sql
-- Query 4: Non-overlapping ranges for different cities

EXPLAIN FORMAT=BRIEF 
    SELECT * FROM listings 
    WHERE
        (city = 'San Francisco' AND bedrooms = 1 AND price >= 1500 AND price < 2500) 
     OR (city = 'San Diego' AND bedrooms = 1 AND price >= 1000 AND price < 1500);

>  -----+------------------------------------------------------+--------------------------

Author: Fixed
